File: app/src/main/java/com/example/papersoccer/papersoccer/AI/EuclideanAI.java
package com.example.papersoccer.papersoccer.AI;

import com.example.papersoccer.papersoccer.AI.Abstraction.IGameAI;
import com.example.papersoccer.papersoccer.GameObjects.GameHandler;
import com.example.papersoccer.papersoccer.GameObjects.Move.PartialMove;
import com.example.papersoccer.papersoccer.GameObjects.Move.PossibleMove;
import com.example.papersoccer.papersoccer.Helpers.MathHelper;

public class EuclideanAI implements IGameAI {

    @Override
    public PartialMove MakeMove(GameHandler gameHandler) {
        PartialMove manhattanMove = null;
        double manhattanDistance = Integer.MAX_VALUE;
        double tempManhattan = 0;

        for (PossibleMove possibleMove : gameHandler.allPossibleMovesFromNode(gameHandler.ballNode))
        {
            tempManhattan = MathHelper.euclideanDistance(possibleMove.newNode.xCord, gameHandler.getOpponent(gameHandler.currentPlayersTurn).goalNode.xCord,
                    possibleMove.newNode.yCord, gameHandler.getOpponent(gameHandler.currentPlayersTurn).goalNode.yCord);
            if (tempManhattan < manhattanDistance)
            {
                manhattanDistance = tempManhattan;
                manhattanMove = new PartialMove(possibleMove.oldNode, possibleMove.newNode, gameHandler.currentPlayersTurn);
            }
        }
        return manhattanMove;
    }
}

File: app/src/main/java/com/ps/simplepapersoccer/ai/euclideanAI/EuclideanAI.kt
package com.ps.simplepapersoccer.ai.euclideanAI

import com.ps.simplepapersoccer.ai.abstraction.IGameAI
import com.ps.simplepapersoccer.gameobjects.game.GameHandler
import com.ps.simplepapersoccer.gameobjects.move.PartialMove
import com.ps.simplepapersoccer.gameobjects.player.AIPlayer
import com.ps.simplepapersoccer.helpers.PathFindingHelper

//Dumb AI to use as test opponent
class EuclideanAI(playerNumber: Int,
                  playerColor: Int,
                  playerName: String = EuclideanAI::class.java.simpleName) : IGameAI, AIPlayer(playerName, playerNumber, playerColor) {

    override suspend fun makeMove(gameHandler: GameHandler): PartialMove {
        var manhattanMove: PartialMove? = null
        var manhattanDistance = Int.MAX_VALUE
        var tempManhattan: Int

        for (possibleMove in gameHandler.gameBoard.allLegalMovesFromBallNode) {
            tempManhattan = PathFindingHelper.findPathGreedyBestFirstSearchBiDirectional(possibleMove.newNode, gameHandler.getOpponent(gameHandler.currentPlayersTurn).goal!!.goalNode()).size
            if (tempManhattan < manhattanDistance) {
                manhattanDistance = tempManhattan
                manhattanMove = PartialMove(possibleMove.oldNode, possibleMove.newNode, gameHandler.gameBoard.currentPlayersTurn)
            }
        }
        return manhattanMove!!
    }
}
